### PR TITLE
Plugin group lifecycle e2e framework tooling and test cases implementation

### PR DESCRIPTION
**Pull request by chandrareddyp**
### What this PR does / why we need it
**What this PR does:**
This PR implements end-to-end test framework tooling and test case for plugin group use cases, it covers below plugin group use cases:

```
tanzu plugin group search
tanzu plugin install all --group <plugin_group_name>
tanzu plugin install --group <plugin_group_name>
tanzu plugin install <specific_plugin_name> --group <plugin_group_name>
```
And this PR covers the negative use cases like, 
```
tanzu plugin install <specific_plugin_name> --group <incorrect_plugin_group_name>
tanzu plugin install <incorrect_specific_plugin_name> --group <incorrect_plugin_group_name>
tanzu plugin install <plugin_name_notExistsThisGroup> --group <plugin_group_name>
```
**Why we need this PR:**
This PR helps ensure the plugin group use cases (positive and negative) always work fine from an end-user perspective, giving confidence to the CLI team.
This test executes as part of the "Tanzu CLI Core E2E Tests" GitHub workflow for every pull request. 

**End-to-End framework tooling:**
This PR uses the existing local central repository tooling (https://github.com/vmware-tanzu/tanzu-cli/tree/main/hack/central-repo) as the central repository and also uses the current plugin groups and plugins for the life cycle tests. The required plugin groups and plugin names (for plugin life cycle tests) are hard-coded in the framework code, created issue https://github.com/vmware-tanzu/tanzu-cli/issues/122 to make it configurable.

### Which issue(s) this PR fixes

Fixes #https://github.com/vmware-tanzu/tanzu-cli/issues/112

### Describe testing done for PR
1. Successfully tested locally:
Here are steps to test the PR locally:
```
❯ make build
build darwin-amd64 CLI with version: v0.1.0-dev
# command-line-arguments
ld: warning: -no_pie is deprecated when targeting new OS versions
ld: warning: non-standard -pagezero_size is deprecated when targeting macOS 13.0 or later
mkdir -p bin
cp /Users/cpamuluri/tkg/tasks/e2e_tests/pluginLifeCycle/tanzu-cli/artifacts/darwin/amd64/cli/core/v0.1.0-dev/tanzu-cli-darwin_amd64 ./bin/tanzu
❯                                                                                                
❯ 
❯ make start-test-central-repo
docker container stop central 2> /dev/null || true
central
if [ ! -d /Users/cpamuluri/tkg/tasks/e2e_tests/pluginLifeCycle/tanzu-cli/hack/central-repo/registry-content ]; then \
                (cd /Users/cpamuluri/tkg/tasks/e2e_tests/pluginLifeCycle/tanzu-cli/hack/central-repo && tar xjf registry-content.bz2 || true;) \
        fi
docker run --rm -d -p 9876:5000 --name central \
                -v /Users/cpamuluri/tkg/tasks/e2e_tests/pluginLifeCycle/tanzu-cli/hack/central-repo/registry-content:/var/lib/registry \
                mirror.gcr.io/library/registry:2
be34a05e5159dc5fd1edbdad503a56531db50e2c0924fe768d2f3e7f6733f6d0
❯ 
❯ 
❯ export TANZU_CLI_E2E_TEST_LOCAL_CENTRAL_REPO_URL=localhost:9876/tanzu-cli/plugins/central:small
❯ 
❯ export PATH=/Users/cpamuluri/tkg/tasks/e2e_tests/pluginLifeCycle/tanzu-cli/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/Library/Apple/usr/bin:/Users/cpamuluri/.fig/bin:/Users/cpamuluri/.local/bin
❯ 
❯ make e2e-cli-plugin-lifecycle-test                                                             
=== RUN   TestPluginLifecycle
Running Suite: PluginLifecycle Suite
====================================
Random Seed: 1679350377
Will run 30 of 30 specs

••••[x] error while executing command:'tanzu plugin source delete FhYvt', error:'error while running 'tanzu plugin source delete FhYvt', stdOut: , stdErr: [x] : cli discovery source not found
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : cli discovery source not found
'
••••[x] error while executing command:'tanzu plugin describe cluster --target iMFBv', error:'error while running 'tanzu plugin describe cluster --target iMFBv', stdOut: , stdErr: [x] : invalid target specified. Please specify correct value of `--target` or `-t` flag from 'kubernetes/k8s/mission-control/tmc'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : invalid target specified. Please specify correct value of `--target` or `-t` flag from 'kubernetes/k8s/mission-control/tmc'
'
•[x] error while executing command:'tanzu plugin describe 4nn74 --target kubernetes', error:'error while running 'tanzu plugin describe 4nn74 --target kubernetes', stdOut: , stdErr: [x] : unable to find plugin '4nn74' for target 'kubernetes'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : unable to find plugin '4nn74' for target 'kubernetes'
'
••••••[x] error while executing command:'tanzu plugin install pinniped-auth --target global --version v9.9.9', error:'error while running 'tanzu plugin install pinniped-auth --target global --version v9.9.9', stdOut: , stdErr: [!] unable to list plugin from discovery 'default_pre_release': invalid target for plugin: global
[x] : unable to find plugin 'pinniped-auth' for target 'global'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[!] unable to list plugin from discovery 'default_pre_release': invalid target for plugin: global
[x] : unable to find plugin 'pinniped-auth' for target 'global'
'
•[x] error while executing command:'tanzu plugin install cluster --target U0vfp --version v9.9.9', error:'error while running 'tanzu plugin install cluster --target U0vfp --version v9.9.9', stdOut: , stdErr: [x] : invalid target specified. Please specify correct value of `--target` or `-t` flag from 'kubernetes/k8s/mission-control/tmc'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : invalid target specified. Please specify correct value of `--target` or `-t` flag from 'kubernetes/k8s/mission-control/tmc'
'
•[x] error while executing command:'tanzu plugin install fOmeD', error:'error while running 'tanzu plugin install fOmeD', stdOut: , stdErr: [x] : unable to find plugin 'fOmeD'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : unable to find plugin 'fOmeD'
'
•[x] error while executing command:'tanzu plugin install cluster --target kubernetes --version v9.9.9492', error:'error while running 'tanzu plugin install cluster --target kubernetes --version v9.9.9492', stdOut: , stdErr: [x] : unable to find plugin 'cluster' for target 'kubernetes'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : unable to find plugin 'cluster' for target 'kubernetes'
'
•••••••••[x] error while executing command:'tanzu plugin install cluster --group teEbk', error:'error while running 'tanzu plugin install cluster --group teEbk', stdOut: , stdErr: [x] : could not find group 'teEbk'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : could not find group 'teEbk'
'
•[x] error while executing command:'tanzu plugin install v1GVJ --group vmware-tkg/v9.9.9', error:'error while running 'tanzu plugin install v1GVJ --group vmware-tkg/v9.9.9', stdOut: , stdErr: [x] : plugin 'v1GVJ' is not part of the group 'vmware-tkg/v9.9.9'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : plugin 'v1GVJ' is not part of the group 'vmware-tkg/v9.9.9'
'
•[0xc0005135a0 0xc0005135b0]
[x] error while executing command:'tanzu plugin install audit --group vmware-tkg/v9.9.9', error:'error while running 'tanzu plugin install audit --group vmware-tkg/v9.9.9', stdOut: , stdErr: [x] : plugin 'audit' is not part of the group 'vmware-tkg/v9.9.9'
%!(EXTRA *exec.ExitError=exit status 1)' stdErr:'[x] : plugin 'audit' is not part of the group 'vmware-tkg/v9.9.9'
'
•
Ran 30 of 30 Specs in 16.814 seconds
SUCCESS! -- 30 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestPluginLifecycle (16.84s)
PASS
coverage: 94.3% of statements
ok      github.com/vmware-tanzu/tanzu-cli/test/e2e/plugin_lifecycle     17.093s coverage: 94.3% of statements
❯ 
```

### Release note

```release-note
Plugin life cycle end-to-end test cases has implemented
```

#### Special notes for your reviewer
This is the second PR after https://github.com/vmware-tanzu/tanzu-cli/pull/104 for the plugin life cycle test cases implementation

### CODE REVIEW

File: test/e2e/framework/framework.go [hunk 1]
```diff
@@ -73,20 +76,27 @@ const (
...
+	TMC            = "tmc"
+	TKG            = "tkg"
+	Kubernetes     = "kubernetes"
+	MissionControl = "mission-control"
 	TestDir        = ".tanzu-cli-e2e"
 	TestPluginsDir = ".e2e-test-plugins"
-	TargetTypeTMC  = "mission-control"
-	TargetTypeK8s  = "kubernetes"
+	TargetTypeTMC  = MissionControl
+	TargetTypeK8s  = Kubernetes
```
Comment (anujc25) [inline test/e2e/framework/framework.go:86]: Why are we defining these constants again when we can get those from `tanzu-plugin-runtime`?

File: test/e2e/plugin_lifecycle/plugin_group_lifecycle_test.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		// Test case: b. install a plugin with incorrect plugin name (which is not exists in given plugin group) and correct group name
+		It("install a plugin with incorrect plugin name (which is not exists in given group) and correct group name", func() {
+			for pg := range pluginGroupToPluginListMap {
+				err := tf.PluginCmd.InstallPluginsFromGroup(framework.RandomString(5), pg)
+				Expect(err).NotTo(BeNil(), "plugin installation should fail as plugin name is not exists in plugin group")
+				break
+			}
+		})
+		// Test case: c. install a plugin which is not exists in a plugin group which is passing as plugin group name, but the plugin name exists in some other plugin group.
+		It("install a plugin with incorrect plugin name (which is not exists in given group but exists in other group) and correct group name", func() {
```
Comment (anujc25) [inline test/e2e/plugin_lifecycle/plugin_group_lifecycle_test.go:150]: nit:
```suggestion
		It("install a plugin with incorrect plugin name (which does not exist in the given group but exists in other group) and correct group name", func() {
```

File: test/e2e/plugin_lifecycle/plugin_group_lifecycle_test.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		It("install a plugin with incorrect plugin group name", func() {
+			for pg := range pluginGroupToPluginListMap {
+				plugins := pluginGroupToPluginListMap[pg]
+				err := tf.PluginCmd.InstallPluginsFromGroup(plugins[0].Name, framework.RandomString(5))
+				Expect(err).NotTo(BeNil(), "plugin installation should fail as plugin group name is incorrect")
+				break
+			}
+		})
+		// Test case: b. install a plugin with incorrect plugin name (which is not exists in given plugin group) and correct group name
+		It("install a plugin with incorrect plugin name (which is not exists in given group) and correct group name", func() {
```
Comment (anujc25) [inline test/e2e/plugin_lifecycle/plugin_group_lifecycle_test.go:142]:
```suggestion
		It("install a plugin with incorrect plugin name (which does not exist in given group) and correct group name", func() {
```

File: test/e2e/plugin_lifecycle/plugin_group_lifecycle_test.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+			pluginsList, err := tf.PluginCmd.ListPlugins()
+			Expect(err).To(BeNil(), "should not get any error for plugin list")
+			Expect(len(pluginsList)).Should(BeNumerically("==", 0), "plugins list should not return any plugins after plugin clean")
+		})
+		// Test case: install a plugin from each plugin group and validate the installation
+		It("install a plugin from each plugin group and validate the plugin installation by 'plugin describe'", func() {
+			for pg := range pluginGroupToPluginListMap {
+				plugins := pluginGroupToPluginListMap[pg]
+				err := tf.PluginCmd.InstallPluginsFromGroup(plugins[0].Name, pg)
+				Expect(err).To(BeNil(), "should not get any error for plugin install from plugin group")
```
Comment (anujc25) [inline test/e2e/plugin_lifecycle/plugin_group_lifecycle_test.go:51]: Highly unlikely but what if a plugin group contains the same named plugin with different targets? 

For example, a plugin group `vmware-tkg/v1.0.0` contains a plugin `name:foo target:global` and a plugin `name:foo target:kubernetes`.

In this case what will happen when a user does `tanzu plugin install foo --group vmware-tkg/v1.0.0`?

cc @marckhouzam @chandrareddyp

File: test/e2e/framework/framework_helper.go [hunk 5]
```diff
@@ -71,3 +84,31 @@ func GetHomeDir() string {
 	}
 	return home
 }
+
+// ExecuteCmdAndBuildJSONOutput is generic function to execute given command and build JSON output and return
+func ExecuteCmdAndBuildJSONOutput[T PluginInfo | PluginGroup | PluginSourceInfo | configapi.ClientConfig | Server | ContextListInfo](cmdExe CmdOps, cmd string) ([]*T, error) {
```
Comment (anujc25) [inline test/e2e/framework/framework_helper.go:89]: 👏

File: test/e2e/framework/framework.go (hunk 1 above)
Comment (chandrareddyp) [reply test/e2e/framework/framework.go]: updated with runtime api

**Review by anujc25 — APPROVED** [review]
Changes LGTM. The test pipeline is failing with the recent changes. Once fixed, we should be good to get this in.

**Pull request merged**